cwltool/singularity.py

```diff
@@ -171,29 +172,43 @@ def get_image(
-            singularityfile_path = dockerfile_path + ".def"
-            # if you do not set APPTAINER_TMPDIR will crash
-            # WARNING: 'nodev' mount option set on /tmp, it could be a
-            #          source of failure during build process
-            # FATAL:   Unable to create build: 'noexec' mount option set on
-            #          /tmp, temporary root filesystem won't be usable at this location
-            with open(dockerfile_path, "w") as dfile:
-                dfile.write(dockerRequirement["dockerFile"])
-
-            singularityfile = SingularityWriter(DockerParser(dockerfile_path).parse()).convert()
-            with open(singularityfile_path, "w") as file:
-                file.write(singularityfile)
-
-            os.environ["APPTAINER_TMPDIR"] = absolute_path
-            singularity_options = ["--fakeroot"] if not shutil.which("proot") else []
-            Client.build(
-                recipe=singularityfile_path,
-                build_folder=absolute_path,
-                sudo=False,
-                options=singularity_options,
-            )
-            found = True
+            if "dockerImageId" in dockerRequirement:
+                image_name = dockerRequirement["dockerImageId"]
+                image_path = os.path.join(absolute_path, image_name)
+                if os.path.exists(image_path):
+                    found = True
+            if found is False:
+                dockerfile_path = os.path.join(absolute_path, "Dockerfile")
+                singularityfile_path = dockerfile_path + ".def"
+                # if you do not set APPTAINER_TMPDIR will crash
+                # WARNING: 'nodev' mount option set on /tmp, it could be a
+                #          source of failure during build process
+                # FATAL:   Unable to create build: 'noexec' mount option set on
+                #          /tmp, temporary root filesystem won't be usable at this location
+                with open(dockerfile_path, "w") as dfile:
+                    dfile.write(dockerRequirement["dockerFile"])
+
+                singularityfile = SingularityWriter(DockerParser(dockerfile_path).parse()).convert()
+                with open(singularityfile_path, "w") as file:
+                    file.write(singularityfile)
+
+                os.environ["APPTAINER_TMPDIR"] = absolute_path
+                singularity_options = ["--fakeroot"] if not shutil.which("proot") else []
+                Client.build(
+                    recipe=singularityfile_path,
+                    build_folder=absolute_path,
+                    sudo=False,
+                    options=singularity_options,
+                )
+                if "dockerImageId" in dockerRequirement:
+                    image_name = dockerRequirement["dockerImageId"]
+                    children = sorted(Path(absolute_path).glob("*.sif"))
+                    image_path = children[0]
```

Comment:
This presumes that there are no `.sif` files in the cache already; why not pass the `image_name` to the `image` parameter of `Client.build` to avoid the need for a rename?

https://singularityhub.github.io/singularity-cli/api/source/spython.main.base.html#spython.main.base.Client.build